rfc/src/rfcs/0008-viewerless-coverage.md

```diff
@@ -0,0 +1,89 @@
+Recently, there have been some issues with the coverage information reported by `cbmc-viewer` (e.g., [#2048](https://github.com/model-checking/kani/issues/2048) or [#1707](https://github.com/model-checking/kani/issues/1707)), forcing us to mark the `--visualize` option as unstable and disable coverage results in the reports (in [#2206](https://github.com/model-checking/kani/pull/2206)).
+
+However, it is possible for Kani to report coverage information without `cbmc-viewer`.
+This only requires a new option in Kani that enables the injection of coverage-oriented checks (similar to `kani::cover` statements) into certain parts of the program under verification.
+The status reported by the injected checks is sufficient to construct a coverage report.
+Note that this gives Kani control on both ends:
+ * **The instrumentation performed** on the program. Eventually, this would allow us to report more precise coverage information, similar to [Rust's instrument-based code coverage](https://doc.rust-lang.org/rustc/instrument-coverage.html).
+ * **The format of the coverage report** to be generated. Similarly, this would allow us to generate coverage data in different formats (see [#1706](https://github.com/model-checking/kani/issues/1706) for GCOV, or [#1777](https://github.com/model-checking/kani/issues/1777) for LCOV). While technically this is also doable from `cbmc-viewer`'s output, development is likely to be faster this way.
+
+Moreover, producing coverage information directly from Kani is likely to increase speed of development and improve testing for coverage features, which translates into faster and more reliable coverage options for users.
+
+## User Experience
+
+For the first version, this experimental feature will be limited to produce a sequence of text lines as follows:
+```
+<file>, <line>, <status>
+```
+where `<status>` is either `COVERED` or `UNCOVERED`.
+
+**Users are not expected to consume this output directly.**
+Instead, coverage data is to be consumed by the [Kani VS Code extension](https://github.com/model-checking/kani-vscode-extension) and displayed as in the following picture:
+
+![Coverage reported on the Kani VS Code Extension](../images/0008/coverage-kani-vscode.png)
+
+How to activate and display coverage information in the extension is out of scope for this RFC.
+We will need to change our documentation (including the tutorial) to use this option if we ever remove `--visualize` in favor of this one.
+
+## Detailed Design
+
+We will add a new unstable `--coverage` verification option to Kani.
+This will prevent `kani-driver` from reporting verification results as usual and instead output the coverage information described above.
+It's likely that this will be done through a new output format that's not exposed to users.
+
+We will also add a new `--coverage-checks` option to `kani-compiler`, which will result in the injection of coverage checks before each Rust statement and terminator[^coverage-experiments].
```

Comment:
How does injecting coverage checks before each statement and terminator compare to Rust's instrumentation-based code coverage? One of the main requests in #1706 and #1777 is to generate coverage information that is consistent with what the compiler generates.

Reply:
I'd like to avoid an option with `cover` and `statement` since that's how we name the macro and it could be confusing at some point. So the option is still `--coverage-checks` in this version, but I'm open to other suggestions.

Reply:
Sure. This is internal to Kani anyways. Not a biggie